criblio/appscope
# Rootdir support in proc utils (#1414)

**This PR**
Extends the cli `utils` package to read the proc directory of a foreign namespace, through the use of a `rootdir` prefix passed on the command line. As a result:
- The `inspect` command now allows container to host/container inspect.
- The `ps` command now supports a `--rootdir` argument to show processes on a host, or sibling container.
- The `inspect` command and the `detach` command now show a helper menu when no argument is provided, to allow the user to choose a process. When the `--rootdir` argument is provided, the helper menu displays scoped processes in the namespace related to the rootdir.

**Testing**
The CLI `proc_utils.go` unit test was updated.
No container-to-host or container-to-container integration tests have been added as it is not yet supported in our test framework. See: https://github.com/criblio/appscope/issues/1311
In place of this, QA instructions should suffice for this change.

**Documentation**
The CLI reference page has been updated to reflect new behavior.

**QA**
- Start two containers like this:
`docker run --rm -it -v /path/to/appscope/:/opt/appscope -v /:/hostfs:ro --privileged ubuntu`
- Run 2 instances of `scope top` on the host 
- Exec into one of the containers and run 2 instances of `scope top`
- Exec into the other container and Try:
```
scope inspect --rootdir /hostfs # should reveal a helper menu with 2 procs
scope inspect --rootdir /hostfs/proc/<pid>/root # where pid is any pid in the sibling container from host perspective
scope inspect --all --rootdir /hostfs # should show output for 2 procs
scope inspect --all --rootdir /hostfs/proc/<pid>/root # where pid is any pid in the sibling container from host perspective
scope ps --rootdir /hostfs # should reveal 2 procs
scope ps --rootdir /hostfs/proc/<pid>/root # where pid is any pid in the sibling container
scope detach --rootdir /hostfs # should reveal a helper menu with 2 procs. FYI detach with rootdir doesn't work yet.
scope detach --rootdir /hostfs/proc/<pid>/root # where pid is any pid in the sibling container from host perspective
```
